### DISCUSSION TITLE
Problem faced during full installation steps

### DISCUSSION THREAD

**Discussion by vipinvijay1986** [egovernments/Digit-Core#42] [Announcements]
FYR  : 
 * Doing full installation of Digit
 * For setup using window 11 and cloud provider is aws  

The problem faced while doing full installation steps step  [5. Terraform Execution: Infrastucture Resoruce Provisioning](https://docs.digit.org/urban/platform/setup-digit/install-on-cloud/on-aws/4.-infra-as-code-terraform#86fd) : 

1. **Creating S3 bucket throw the error** : While doing *terraform apply*  creating s3 bucket does not happen, it fails to connects to aws profile. 
   <u>Possible reason is</u> : we have not defined aws profile to be used while creating s3 bucket. 
2. **Creating Infra setup** : While doing the infra setup it fails as the aws profile is different from previously created steps.  ( Profile not found with name "digit-infra-aws"). It would be better if we ask user to provide profile detail in providers.tf.
 

3. **While destroying the s3 bucket**  : 
    ```
    Error: Instance cannot be destroyed

    on main.tf line 5:
    5: resource "aws_s3_bucket" "terraform_state" {

    Resource aws_s3_bucket.terraform_state has lifecycle.prevent_destroy set, butthe plan calls for this resource to be destroyed. To avoid this error andcontinue with the plan, either disable lifecycle.prevent_destroy or reduce thescope of the plan using the -target flag.
    ```

4. **while destroying the cluster** : execution stops with following line :  
    ```
    module.eks.aws_security_group.cluster[0]: Destroying... [id=sg-03c7ecac0afcef9a1]module.eks.aws_security_group.cluster[0]: Destruction complete after 1s
    module.eks.aws_iam_role.cluster[0]: Destruction complete after 1s

    Error: final_snapshot_identifier is required when skip_final_snapshot is false
    ```
    After this I checked aws console and it was having following state : 
     * EC2 instances are in terminated state. 
     * In EC2 --> Volumes -->vol-0b825d77c58b5bc33 , volume was not deleted. 
     * Network & Secuirty : 
       * Security Groups  exist  ( not able to delete also )
       * Network interfaces: 1 network interface exist.  ( not able to delete also )

Please let me know if any other things to be checked on aws console. As it might create an extra bill amount for me.

**Comment by subhashini-egov** [comment]
@vipinvijay1986 - 
1. S3 buckets have to be created outside of the installer. Only the names of the buckets have to be mentioned here in the terraform template.
2. In providers.tf inside the sample-aws folder, the AWS profile has to match the AWS profile that's been created/used to access the infra. Check the ~/.aws/credentials file and make sure to add the profile name used there in the providers.tf file. Else, remove the profile line and it will still work if AWS_PROFILE env variable is set.
3. Some resources are created outside of terraform. For eg. S3 buckets as well as ELBs which are created during deploy time in the nginx ingress controller. The ELB is outside the purview of terraform but is tied to the VPC. Therefore VPC deletion will fail because it has components depending on it. The way to solve this is:

```kubectl get svc -n <namespace>```

The namespace can be one of `jenkins` or `egov`. Then, go ahead and delete the ingress controller deployment as follows:

```kubectl delete deployment nginx-ingress-controller -n <namespace>```
This will get rid of the ELBs. After this, go ahead and run ```terraform destroy``` again.
4. I think that is a RDS issue. Please go into the management console and delete the RDS instance manually. Re-run terraform destroy again. 

Please try this out. Hopefully, this will release all the resources provisioned by terraform.
